vmdb/app/models/mixins/retirement_mixin.rb

```diff
@@ -80,40 +84,51 @@ def retire(options={})
   def retirement_check
     return if self.retired?
 
-    if !self.retirement_warned? && self.retirement_warning_due?
+    if !retirement_warned? && retirement_warning_due?
       begin
         self.retirement_last_warn = Time.now.utc
-        self.save
+        save
         raise_retirement_event(retire_warn_event_name)
       rescue => err
         $log.log_backtrace(err)
       end
     end
 
-    self.retire_now if self.retirement_due?
+    retire_now if retirement_due?
   end
 
   def retire_now
     log_prefix = "MIQ(#{retirement_object_title}#retire_now)"
-    unless self.retired
-      $log.info("#{log_prefix} Retiring [#{self.name}]")
-      before_retirement
-      self.retires_on = Date.today
-      self.retired    = true
-      self.save
-      message = "#{retirement_object_title}: [#{self.name}], Retires On Date: [#{self.retires_on_date}], has been retired"
-      $log.info("#{log_prefix} #{message}")
-      raise_audit_event(retired_event_name, message)
-    else
+
+    if retired
       return if retired_validated?
       $log.info("#{log_prefix} #{retirement_object_title}: [#{self.name}], Retires On Date: [#{self.retires_on_date}], was previously retired, but currently #{retired_invalid_reason}")
+    else
+      event_name = "request_#{retirement_event_prefix}_retire"
+      $log.info("#{log_prefix} calling #{event_name}")
+      begin
+        raise_retirement_event(event_name)
+      rescue => err
+        $log.log_backtrace(err)
+      end
     end
+  end
 
-    begin
-      raise_retirement_event(retired_event_name)
-    rescue => err
-      $log.log_backtrace(err)
-    end
+  def finish_retirement
+    return if self.retired?
+
+    $log.info("Finishing Retirement for [#{name}]")
+    update_attributes(:retires_on => Date.today, :retired => true, :retirement_state => "retired")
+    message = "#{self.class.base_model.name}: [#{name}], Retires On Date: [#{retires_on}], has been retired"
+    $log.info("Calling audit event for: #{message} ")
+    raise_audit_event(retired_event_name, message)
+    $log.info("Called audit event for: #{message} ")
+  end
+
+  def start_retirement
+    return if self.retired?
```

Comment:
The automate methods are checking the retirement state before calling this method but I think you want to raise an error here if it gets called and the resource is already retired.  Otherwise we are giving the false impression that we are ok to proceed.  Imagine others writing their own retirement logic in automate and not being so diligent in checking the retirement state first.